group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -195,6 +196,11 @@ public OffsetMetadataManager build() {
      */
     private final TimelineHashMap<Long, Offsets> pendingTransactionalOffsets;
 
+    /**
+     * The open transactions (producer ids) keyed by group.
+     */
+    private final TimelineHashMap<String, TimelineHashSet<Long>> openTransactionsByGroup;
```

Comment:
I follow an approach similar to the one used in the scala implementation. We basically store the open producer ids per group and use them to search the corresponding partitions in `pendingTransactionalOffsets`. An alternative way would be to directly store the open transactions by group, topic and partition but this would use more space. Classic runtime vs space tradeoff.

Reply:
just for my understanding though -- for every producer ID for a group we return, we will have an offset in pendingTransactionalOffsets, and every producer ID in pendingTransactionalOffsets will have its group in openTransactionsByGroup.

I can't imagine we would have a case where something is in one map but not the other. (I can imagine that the group is not there at all or the topic partition is not there though)

Reply:
Your understanding is correct.

Reply:
Sorry -- I left a lot of words here. Is there a time where we take advantage of the data structure? Ie using the epoch/rolling back? I didn't see any in this PR.

Reply:
yeah, it is implicit. the write operation (the transactional offset commit) could for instance fail. the state is rolled back in this case.

Reply:
Ah its the snapshot registry and revertLastWrittenOffset method that does this. I had forgotten. But makes sense now.